File: pattern-01-array-string/richest_customer_wealth.cpp
// Problem: Richest Customer Wealth
// Link: https://leetcode.com/problems/richest-customer-wealth/
// Pattern: Matrix Traversal
// Difficulty: Easy
// Date: 2026-05-12

// Approach: Traverse each row of the 2D array, calculate the sum of all
// bank balances for each customer, and keep track of the maximum sum found.
// Time: O(m × n) | Space: O(1)

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int maximumWealth(vector<vector<int>>& accounts) {
        int maxWealth = 0;
        int rows = accounts.size();
        int columns = accounts[0].size();

        for (int i = 0; i < rows; i++) {
            int wealth = 0;

            for (int j = 0; j < columns; j++) {
                wealth += accounts[i][j];
            }

            maxWealth = max(maxWealth, wealth);
        }

        return maxWealth;
    }
};

// Local testing with manual input
int main() {
    int m, n;

    cout << "Enter number of customers (rows): ";
    cin >> m;

    cout << "Enter number of bank accounts (columns): ";
    cin >> n;

    vector<vector<int>> accounts(m, vector<int>(n));

    cout << "Enter the bank balances for each customer:\n";
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            cin >> accounts[i][j];
        }
    }

    Solution sol;
    cout << "Maximum Wealth: "
         << sol.maximumWealth(accounts) << endl;

    return 0;
}

File: pattern-01-array-string/richest_customer_wealth.py
# Problem: Richest Customer Wealth
# Link: https://leetcode.com/problems/richest-customer-wealth/
# Pattern: Matrix Traversal
# Difficulty: Easy
# Date: 2026-05-12

# Approach: Iterate through each row (customer), calculate the sum of all
# account balances, and keep track of the maximum row sum.
# Time: O(m × n) | Space: O(1)

from typing import List

class Solution:
    def maximumWealth(self, accounts: List[List[int]]) -> int:
        max_row_sum = float('-inf')

        for row in accounts:
            current_row_sum = 0

            for amount in row:
                current_row_sum += amount

            if current_row_sum > max_row_sum:
                max_row_sum = current_row_sum

        return max_row_sum


# Local testing with manual input
if __name__ == "__main__":
    accounts = eval(input("Enter accounts (e.g. [[1,2,3],[3,2,1]]): "))
    sol = Solution()
    print("Maximum Wealth:", sol.maximumWealth(accounts))